src/main.cpp

```diff
@@ -1,224 +1,115 @@
-    // Destroy all objects created by the FBX SDK.
-    DestroySdkObjects(lSdkManager, lResult);
-
-    return 0;
-}
-
-bool CreateScene(FbxScene *pScene, map<string, string> setting) {
-    FbxNode *lNurbs = CreateNurbs(pScene, "Face");
-    // Build the node tree.
-    FbxNode *lRootNode = pScene->GetRootNode();
-    lRootNode->AddChild(lNurbs);
-    MapShapesOnNurbs(pScene, lNurbs, setting);
-    AnimateNurbs(lNurbs, pScene, setting);
-    return true;
-}
-
-// Create a sphere.
-FbxNode *CreateNurbs(FbxScene *pScene, const char *pName) {
-    FbxNurbs *lNurbs = FbxNurbs::Create(pScene, pName);
-
-    // Set nurbs properties.
-    lNurbs->SetOrder(4, 4);
-    lNurbs->SetStep(2, 2);
-    lNurbs->InitControlPoints(8, FbxNurbs::ePeriodic, 7, FbxNurbs::eOpen);
-
-    double lUKnotVector[] = {-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0};
-    memcpy(lNurbs->GetUKnotVector(), lUKnotVector, lNurbs->GetUKnotCount() * sizeof(double));
-
-    double lVKnotVector[] = {0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 3.0, 4.0, 4.0, 4.0, 4.0};
-    memcpy(lNurbs->GetVKnotVector(), lVKnotVector, lNurbs->GetVKnotCount() * sizeof(double));
-
-    FbxVector4 *lVector4 = lNurbs->GetControlPoints();
-    int i, j;
-    double lScale = 20.0;
-    double lPi = 3.14159;
-    double lYAngle[] = {90.0, 90.0, 52.0, 0.0, -52.0, -90.0, -90.0};
-    double lRadius[] = {0.0, 0.283, 0.872, 1.226, 0.872, 0.283, 0.0};
-
-    for (i = 0; i < 7; i++) {
-        for (j = 0; j < 8; j++) {
-            double lX = lScale * lRadius[i] * cos(lPi / 4 * j);
-            double lY = lScale * sin(2 * lPi / 360 * lYAngle[i]);
-            double lZ = lScale * lRadius[i] * sin(lPi / 4 * j);
-            double lWeight = 1.0;
-
-            lVector4[8 * i + j].Set(lX, lY, lZ, lWeight);
+    // アニメーションを記録するメッシュの作成
+    FbxMesh *lMesh = CreateSquareMesh(lScene);
+    FbxNode *lMeshNode = FbxNode::Create(lScene, "VMDshapeAnimation");
+    lMeshNode->SetNodeAttribute(lMesh);
+    lScene->GetRootNode()->AddChild(lMeshNode);
+
+    // そのメッシュにおいて、シェイプキーとそれに対応する AnimationCurve を作成
+    ConfigureBlendShapeDeformer(lMesh, lAnimLayer, shapekeynames);
+    unordered_map<string, FbxAnimCurve *> shapecurvemap = CreateShapeCurveMap(lMesh, lAnimLayer);
+
+    // フレーム毎にシェイプキーに対応する AnimationCurve を取得してアニメーションを記録
+    for (const auto &frame : skindata)
+    {
+        string shapekeyname = sjis_to_utf8(string(frame.SkinName));
+        auto it = shapecurvemap.find(shapekeyname);
+        if (it != shapecurvemap.end())
+        {
+            FbxAnimCurve *lCurve = it->second;
+            if (lCurve)
+            {
+                lCurve->KeyModifyBegin();
+
+                // 表情の1フレーム分を登録、MMDでのシェイプキー値は 0~1 なので注意
+                FbxTime lTime;
+                lTime.SetFrame(frame.FrameNo, FbxTime::eFrames30); // MMD は 30fps
+                int keyindex = lCurve->KeyAdd(lTime);
+                lCurve->KeySet(keyindex, lTime, frame.Weight * 100.0, FbxAnimCurveDef::eInterpolationLinear);
```

Comment:
```suggestion
                // 表情の1フレーム分を登録
                FbxTime lTime;
                lTime.SetFrame(frame.FrameNo, FbxTime::eFrames30); // MMD は 30fps
                int keyindex = lCurve->KeyAdd(lTime);
                // MMDでのシェイプキー値は 0~1 なので、fbxに合わせて100倍する
                lCurve->KeySet(keyindex, lTime, frame.Weight * 100.0, FbxAnimCurveDef::eInterpolationLinear);
```
コメントは近くに書いて、処理に関して説明してあげる方が丁寧で良いです。

Reply:
了解です、修正します。